### DISCUSSION TITLE
Best practices to load custom resources

### DISCUSSION THREAD

**Discussion by thomas-aw** [symfony/symfony#42717] [Q&A] [answered]
Hi all ! 
I have a 4.4 symfony app and i want to manage some configuration files with the same structure. Configuration will be located at : config/my_awesome_conf /*.yaml and can contains some files with structure :
```
title: a random title
formula: x * y
```

I wanted to know the best practice to load all these configuration file at kernel boot to use it later in some services. Maybe an extension ? 
Many thanks

**Comment by faizanakram99** [comment]
`Kernel::configureContainer()` is the method you're looking for. It is used to import resources, services, etc from different locations.
As an example (This example loads both yaml and php configs)

```php
protected function configureContainer(ContainerConfigurator $container): void
{
    $container->import('../config/{packages}/*.{php,yaml}');
    $container->import(
        "../config/{packages}/{$this->environment}/*.{php,yaml}"
    );
    $container->import(
        '../config/my_awesome_conf/*.{php,yaml}',
        null,
        'not_found'
    );

    if (\is_file(\dirname(__DIR__).'/config/services.yaml')) {
        $container->import('../config/services.yaml');
        $container->import(
            "../config/{services}_{$this->environment}.yaml"
        );
    } else {
        $container->import('../config/{services}.php');
        $container->import(
            "../config/{services}_{$this->environment}.php"
        );
    }
}
```

**Reply by thomas-aw** [reply]
I have 4.4 and the signature of configureContainer is different 

`
    abstract protected function configureContainer(ContainerBuilder $container, LoaderInterface $loader);`

I tried `$container->load` but it has to have a specific configuration tree with root but my files don't have root.

But i can create a dedicated cache with ConfigCache in configureContainer but it's linked with container cache so it doesn't detect change on my files.

**Reply by faizanakram99** [reply]
What configs are in that awesome config file ?

You should be able to simply import it in `config.yaml` file

```yaml
imports:
    # glob expressions are also supported to load multiple files
    - { resource: 'my_awesome_conf/*.yaml ' }
```

**Reply by thomas-aw** [reply]
Actually these file are use to configure a list.
Example :
```yaml
# myfirstlist.yaml
name: My first list
link:
    click: '' # On line click
    new: '' # On add a new line
entity: Article
```
```yaml
# mysecond.yaml
name: My second list
link:
    click: '' # On line click
    new: '' # On add a new line
entity: Book
```

My goal is to put these configuration in cache and get it after in some service.
Actually, it's almost the same as routing with yaml loader but i don't where to load it

**Comment by lyrixx** [comment, marked as answer]
Hello,

## You have many options here

First you need to decide if you want to validate files or not.
* If you want to validate files, you'll have to use the Config component and to [create a tree builder](https://symfony.com/doc/current/components/config.html)
* If you don't want, you can use [YAML](https://symfony.com/doc/current/components/yaml.html) component directly

Then,

* you can inline everything directly in the `kernel::build()` method
* Or you can create an extension, and register it in `kernel::build()`

---

## My advices

For what it worth:

* If I have only one file, I would not create a config tree builder, useless
* If I have many files, and many people working on that app, I may create a config tree builder

---

## Examples:

### With no validation and everything wired manually

```php
    protected function build(ContainerBuilder $container)
    {
        $files = (new Finder())
            ->in(__DIR__ . '/../config/my-config')
            ->files()
            ->name('*.yaml')
        ;

        foreach ($files as $file) {
            $config = Yaml::parseFile($file);
            dump($config);
        }

        die;
    }
```

### With validation with a tree builder and everything wired manually

```php
    protected function build(ContainerBuilder $container)
    {
        $files = (new Finder())
            ->in(__DIR__ . '/../config/my-config')
            ->files()
            ->name('*.yaml')
        ;

        $configuration = new class implements ConfigurationInterface {
            public function getConfigTreeBuilder() {
                $treeBuilder = new TreeBuilder('my_config');
                $rootNode = $treeBuilder->getRootNode();

                $rootNode
                    ->children()
                        ->scalarNode('name')
                            ->cannotBeEmpty()
                        ->end()
                    ->end()
                ;

                return $treeBuilder;
            }
        };

        foreach ($files as $file) {
            $config = (new Processor())->processConfiguration($configuration, [Yaml::parseFile($file)]);
            $container->addResource(new FileResource((string) $file));
            dump($config);
        }

        die;
    }
```

In this case files have the following shape:

```yaml
name: name 001
```

### With validation with a tree builder via an extension

```php
    protected function build(ContainerBuilder $container)
    {
        $container->registerExtension(new class extends Extension {
            public function load(array $configs, ContainerBuilder $container)
            {
                $configuration = new class implements ConfigurationInterface {
                    public function getConfigTreeBuilder() {
                        $treeBuilder = new TreeBuilder('my_config');
                        $rootNode = $treeBuilder->getRootNode();

                        $rootNode
                            ->prototype('array')
                                ->children()
                                    ->scalarNode('name')
                                        ->cannotBeEmpty()
                                    ->end()
                                ->end()
                            ->end()
                        ;

                        return $treeBuilder;
                    }
                };

                $config = $this->processConfiguration($configuration, $configs);
                dump($config);
                die;
            }

            public function getAlias()
            {
                return 'my_config';
            }
        });
    }
```

In this case, the you need to edit `Kernel::configureContainer()` (because you want another folder) and add:

```php
$container->import('../config/packages/my-config/*.yaml');
```

Files have the following shape:

```yaml
my_config:
    foo:
        name: name 001
```

**Reply by thomas-aw** [reply]
Thanks @lyrixx, the registerExtension looks nice. But i cannot use treebuilder cause my file structure doesn't have any Root

**Reply by lyrixx** [reply]
no problem :) Don't forget to mark the question *answered* :)

**Reply by thomas-aw** [reply]
@lyrixx A last question, you dump the parsed configuration, but how to add in container cache ? 

Thanks ;)

**Reply by thomas-aw** [reply]
I will use my own cache config, hope it's a good solution ;)
